Std/Data/Array/Lemmas.lean

```diff
@@ -844,3 +848,61 @@ instance [DecidableEq α] (a : α) (as : Array α) : Decidable (a ∈ as) :=
 /-! ### erase -/
 
 @[simp] proof_wanted erase_data [BEq α] {l : Array α} {a : α} : (l.erase a).data = l.data.erase a
+
+/-! ### swap -/
+
+@[simp] theorem get_swap_right (a : Array α) {i j : Fin a.size} :
+    (a.swap i j)[j.val]'(a.size_swap .. |>.symm ▸ j.isLt) = a[i] :=
+  by simp [swap, fin_cast_val]
+
+@[simp] theorem get_swap_left (a : Array α) {i j : Fin a.size} :
+    (a.swap i j)[i.val]'(a.size_swap .. |>.symm ▸ i.isLt) = a[j] :=
+  if he : ((Array.size_set _ _ _).symm ▸ j).val = i.val then by
+    simp [←he, fin_cast_val]
+  else by
+    apply Eq.trans
+    · apply Array.get_set_ne
+      · simp
+      · assumption
+    · simp [get_set_ne]
+
+@[simp] theorem get_swap_of_ne (a : Array α) {i j : Fin a.size} (hp : p < a.size)
+    (hi : p ≠ i) (hj : p ≠ j) : (a.swap i j)[p]'(a.size_swap .. |>.symm ▸ hp) = a[p] := by
+  apply Eq.trans
+  · have : ((a.size_set i (a.get j)).symm ▸ j).val = j.val := by simp only [fin_cast_val]
+    apply Array.get_set_ne
+    · simp [this]
+      apply Ne.symm
+      · assumption
+  · apply Array.get_set_ne
+    · apply Ne.symm
+      · assumption
+
+theorem get_swap (a : Array α) (i j : Fin a.size) (k : Nat) (hk: k < a.size) :
+    (a.swap i j)[k]'(by simp_all) = if k = i then a[j] else if k = j then a[i]  else a[k] := by
+  split
+  · simp_all
+  · split <;> simp_all
+
+theorem get_swap' (a : Array α) (i j : Fin a.size) (k : Nat) (hk' : k < (a.swap i j).size) :
+    (a.swap i j)[k] = if k = i then a[j] else if k = j then a[i] else a[k]'(by simp_all) := by
+  apply get_swap
+
+@[simp] theorem swap_swap (a : Array α) {i j : Fin a.size} :
+    (a.swap i j).swap ⟨i.1, (a.size_swap ..).symm ▸ i.2⟩ ⟨j.1, (a.size_swap ..).symm ▸ j.2⟩ = a := by
```

Comment:
I use `⟨i.1, (a.size_swap ..).symm ▸ i.2⟩` instead of `Fin.cast  (a.size_swap ..).symm i` because the former in definitionally equivalent but I found that it makes my life easier in actual usage (I dont have an additional trivial theorem proving that they are equal if I an in a context that expects the former)

Reply:
No problem. If `simp` can still cope that I think this is fine.